README.md
OpenMensa Parsers
=================

[OpenMensa](http://openmensa.org) is a free canteen database. It depends on external parsers that provide the meal information for the specific canteens.

This repository contains a large collection of parsers for different canteens all over Germany - mostly university canteens provided by student unions.

Before you continue you may want to read the [feed documentation](http://doc.openmensa.org/feed/v2/) describing the exchange format between parsers and OpenMensa.


## Contribute

**Corrections Wellcome**: As I do not use most of the parsers myself it is likely that I miss some parsing issues. Feel free to report an issue or even better provide a pull request.

**Hosting Provided**: Your canteen is missing? You could write a parser for your canteen (in Python)? But you do not know where to host the parser? I can host your parser at omfeeds.devtation.de. Please provide a PR with the new parser.


## Overall Structure

The parsers itself are independent. But there is a small framework handling common tasks like URL parsing and output generation.

Each provider has it's own Python module. A provider represents a collection of canteens which are organisationally dependent and therefore can be parsed by the same process. The module itself has to implement a `parse_url(canteenidentifier, today=False)` method. This method is called to parse and return the feed for a specific canteen. What the `canteenidentifier` is exactly is up to the provider - mostly they are URL parts or URL suffixes.

The [config.py](blob/master/config.py) contains a list of all known providers and it's canteens (plus the `canteenidentifier` that is passed to the `parse_url` method). The structure is hopefully self explaining. If not, please open an issue.


## Common implementation details of providers

At the moment all providers using [PyOpenMensa](https://pypi.python.org/pypi/pyopenmensa) ([documentation](http://pyom.devtation.de), [repo](https://github.com/mswart/pyopenmensa)) to generate the XML feed and for some help for the parsing itself.

As many meal information are only available online as HTML, [Beautiful Soup 4](http://www.crummy.com/software/BeautifulSoup/) is used as a robust but easy to use HTML parser.


## Get started

1. Clone the source code

        git clone --recurse-submodules git://github.com/mswart/openmensa-parsers

2. Install the dependecies:
   * [Python 3](https://www.python.org/)
   * [Beautiful Soup 4](http://www.crummy.com/software/BeautifulSoup/) - needed for most parsers/providers.

3. Try some parsers

        python3 parse.py magdeburg ovgu-unten

   general:

        python3 parse.py <provider name> <canteen name>


## Tips for adding a new provider

1. Search where the meal information are accessible online. JSON or CSV downloads are mostly the best, HTML sites are also possible, but PDF is tricky.
2. Maybe take some look on other parsers how they solve the problem / which libraries they use.
3. Create the new provider
4. Register your provider with its canteens in `config.py`
5. Submit a PR
6. Wait until the PR is reviewed and merged
6. Register the new canteens on openmensa with the feed from `http://omfeeds.devtation.de/<provider identifier>/<canteen identifier>.xml` and (optional) today feed from `http://omfeeds.devtation.de/<provider identifier>/<canteen identifier>/today.xml`


## Further questions

* Read the [OpenMensa Documentation](http://doc.openmensa.org)
* Ask at the OpenMensa IRC channel `freenode#openmensa`
* The [add howto for developers of new parser](https://github.com/mswart/openmensa-parsers/issues/2) issue may be helpful.
* Open an issue



debian/changelog
openmensa-parsers (1.1) trusty; urgency=medium

  * Upgrade to trusty

 -- Malte Swart <packagestdevtation.de>  Tue, 10 Nov 2015 13:57:15 +0100

openmensa-parsers (1.0) precise; urgency=low

  * Create debian package to improve deploy steps

 -- Malte Swart <packages@devtation.de>  Thu, 10 Oct 2013 21:17:43 +0200


halle.py
#!python3
import re
import datetime

from utils import Parser, EsaySource, Source


class Canteen(EsaySource):
    def __init__(self, *args, location, needed_title, not_halle=False):
        super(Canteen, self).__init__(*args)
        self.location = location
        self.needed_title = needed_title
        self.not_halle = not_halle

    def parse_data(self, **kwargs):
        kwargs['selected_locations[]'] = self.location
        data = self.parse_remote('https://www.meine-mensa.de/speiseplan_iframe',
                                 args=kwargs)
        table = data.find('table', 'speiseplan')

        date = None
        pos = 0
        for tr in table.tbody.find_all('tr', recursive=False):
            if pos == 0:
                if self.needed_title:
                    assert self.needed_title in tr.text, tr.text
                else:
                    print(tr.text)
                pos += 1
                continue
            elif pos == 1:
                pos += 1
                continue
            if 'break' in tr.attrs.get('class', []):
                date = list(tr.find_all('td'))[1].text
                continue
            if 'empty_cell' in tr.attrs.get('class', []):
                continue
            tds = list(tr.find_all('td', recursive=False))
            category = tds[1].find('span', attrs={'class': 'npsble'}).text.strip() or 'Hauptessen'
            name = tds[2].find('img').attrs['alt']
            prices = {'student': tds[3].text, 'employee': tds[4].text, 'other': tds[5].text}
            self.feed.addMeal(date, category, name, prices=prices)

    def extract_metadata(self):
        url_template = 'http://www.studentenwerk-halle.de/hochschulgastronomie/{}/{}/'
        if self.not_halle:
            url = url_template.format(self.name, 'mensa-' + self.name)
        else:
            name = self.name
            if 'mensa' not in name:
                name = 'mensa-' + name
            url = url_template.format(self.parser.local_name, name)
        document = self.parse_remote(url)
        attachment = document.find(id='attachContact')

        canteen = self.feed
        canteen.name = document.find('li', attrs={'class': 'current'}).text
        canteen.availability = 'public'
        canteen.address = attachment.find(attrs={'class': 'address'}).text + ', ' + attachment.find(attrs={'class': 'city'}).text
        canteen.city = re.search('\d{5}\s+(?P<city>.+)', canteen.address).group('city')
        canteen.phone = attachment.find(attrs={'class': 'fon'}).text.split(':')[1].strip()
        #email_span = attachment.find(attrs={'class': 'email'})
        #if email_span:
        #    canteen.email = email_span.text

        script = document.find(id='attachMap').find('script').find_all(text=True)[0]
        match = re.search('\[\s*(?P<long>\d+\.\d+)\s*,\s*(?P<lat>\d+\.\d+)\s*]', script)
        canteen.location(match.group('lat'), match.group('long'))

    @Source.today_feed
    def today(self, request):
        day = datetime.datetime.now()
        self.parse_data(day=day.day, month=day.month, year=day.year)
        return self.feed.toXMLFeed()

    @Source.feed(name='thisWeek', priority=1, hour='8', retry='2 60')
    def thisWeek(self, request):
        day = datetime.datetime.now().isocalendar()
        self.parse_data(week=day[1], year=day[0])
        return self.feed.toXMLFeed()

    @Source.feed(name='nextWeek', priority=2, hour='9')
    def nextWeek(self, request):
        day = (datetime.datetime.now() + 7 * datetime.date.resolution).isocalendar()
        self.parse_data(week=day[1], year=day[0])
        return self.feed.toXMLFeed()


parser = Parser(name='halle', version=1.0)
Canteen('harzmensa', parser, location=3, needed_title='Harzmensa')
Canteen('weinbergmensa', parser, location=5, needed_title='Weinbergmensa')
#Canteen('cafebar-weinberg', parser, location=, needed_title='')
Canteen('tulpe', parser, location=10, needed_title='Mensa Tulpe')
Canteen('heidemensa', parser, location=17, needed_title='Heidemensa')
Canteen('burg', parser, location=12, needed_title='Mensa Burg')
Canteen('neuwerk', parser, location=9, needed_title='Neuwerk')
Canteen('franckesche-stiftungen', parser, location=14, needed_title='Franckesche Stiftungen')

#merseburg = parser.sub('merseburg')
Canteen('merseburg', parser, location=16, needed_title='Mensa Merseburg', not_halle=True)
#Canteen('cafebar-merseburg', merseburg, location=, needed_title=)

#dessau = parser.sub('dessau')
Canteen('dessau', parser, location=13, needed_title='Mensa Dessau', not_halle=True)

koethen = parser.sub('koethen')
Canteen('fasanerieallee', koethen, location=7, needed_title='Mensa Köthen')
#Canteen('lohmannstrasse', koethen, location=, needed_title=None)

#bernburg = parser.sub('bernburg')
Canteen('bernburg', parser, location=8, needed_title='Mensa Bernburg', not_halle=True)
